### DISCUSSION TITLE
Compactor warning: skipped partial block when updating bucket index

### DISCUSSION THREAD

**Discussion by KevinDW-Fluxys** [grafana/mimir#5432] [Help and support] [answered]
Hi,

We notice that every 15 minutes our compactor throws a lot of these warnings (more than 10 000 log entries every 15 min): 
`caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H0WNS17PDCC815QF60TWF68Q``
When i look at for example this block in storage, i can see that it only contains a deletion-mark.json and an empty chunks folder. 
I can also see that the deletion-mark.json file gets deleted every 15 minutes and then recreated.

I can also see that for this tenant (smoke-test) this warning occurs for 118 times, while we have 135 blocks on storage. In the bucket-index there are 17 blocks, so it checks out that these are the blocks not in the index.

What should i do to clean this up? Should i manually delete these blocks from storage?

Also: What would have caused this?

**Comment by aknuds1** [comment]
Which version of Mimir are you using? Can you find messages like "started blocks cleanup and maintenance" or "cleaned up partial blocks" in the logs? What about messages like "error deleting partial block marked for deletion" or "deleted partial block marked for deletion"?

What's the value of the `-compactor.partial-block-deletion-delay` flag? Can you find the message "stale partial block found: marking block for deletion" in your logs?

**Reply by KevinDW-Fluxys** [reply]
We are using 2.9.0

I can find "started blocks cleanup and maintenance" (one for each tenant on each cleanup run) but no "cleaned up partial blocks".

There are also no entries for "error deleting partial block marked for deletion", but a whole lot of "deleted partial block marked for deletion"

Not sure if relevant but i'm also noticing quite some "stale partial block found: marking block for deletion"

**Comment by aknuds1** [comment]
That you find "deleted partial block marked for deletion" looks promising, it ought to mean that partial blocks get deleted. The question is why eventually partial blocks are _not_ deleted. Do you see the log message "error deleting partial block marked for deletion"?

Which other log messages do you see for the block `01H0WNS17PDCC815QF60TWF68Q` (that apparently isn't getting cleaned up)? Can you see either "error deleting partial block marked for deletion" or "deleted partial block marked for deletion" for it?

**Reply by KevinDW-Fluxys** [reply]
There are no "error deleting partial block marked for deletion" messages. The messages i get for a block (i took another block in this example) are:

`ts=2023-07-06T07:25:41.595521673Z caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T07:25:44.173729577Z caller=blocks_cleaner.go:508 level=info component=cleaner user=smoke-test msg="deleted partial block marked for deletion" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T07:42:02.365862598Z caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T07:42:11.068017702Z caller=block.go:170 level=info component=cleaner user=smoke-test msg="block has been marked for deletion" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T07:58:22.198769703Z caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T07:58:25.461106169Z caller=blocks_cleaner.go:508 level=info component=cleaner user=smoke-test msg="deleted partial block marked for deletion" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T08:14:43.708426019Z caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T08:14:50.287005159Z caller=block.go:170 level=info component=cleaner user=smoke-test msg="block has been marked for deletion" block=01H2B1QS3AZ1X4S2RSCA6E49SW
ts=2023-07-06T08:31:02.42020993Z caller=updater.go:112 level=warn component=cleaner user=smoke-test msg="skipped partial block when updating bucket index" block=01H2B1QS3AZ1X4S2RSCA6E49SW`

**Reply by KevinDW-Fluxys** [reply]
on my storage i can also see the deletion-mark being deleted and recreated every time, but the folders stay there and remain empty.

**Comment by aknuds1** [comment]
Assuming it's safe to delete the `01H0WNS17PDCC815QF60TWF68Q` block, could you delete it manually and see if it gets recreated? 

What kind of object storage back-end are you using?

**Reply by KevinDW-Fluxys** [reply]
When i delete a block folder manually, it remains deleted. It seems that the folder does not get deleted by the compactor. Could this be a permissions issue?
We are using Azure Blob Storage, which should not be too exotic I assume?

**Comment by aknuds1** [comment]
I have the suspicion that your object storage backend retains folders after they are emptied. I vaguely recall seeing this issue in the past, with some exotic storage backend.

**Comment by aknuds1** [comment]
> We are using Azure Blob Storage, which should not be too exotic I assume?

Interesting. I will have to check with the team about whether the problem of object storages keeping empty folders behind is still a problem, and whether this problem is indeed present with Azure Blob Storage.

In the meantime, can you tell me whether you are able to (manually) create empty folders in Azure Blob Storage, or whether an empty folder remains after you remove its files? I'm pretty certain empty folders can't exist in neither AWS S3 nor GCS, as they don't have the folder concept (just file paths).

**Reply by KevinDW-Fluxys** [reply]
Yes, I can create an empty folder, and the folders also remain when all the files in it are cleared. 

It would indeed make sense if the cleanup mechanism is detecting these empty folders, marking them as to delete with a delete-mark file, deleting said file and expecting the folder to disappear as well.

**Reply by aknuds1** [reply]
@KevinDW-Fluxys for reference, how do you technically make en empty folder in ABS?

**Reply by KevinDW-Fluxys** [reply]
I did this using the azure storage explorer, its a gui for azure storage

**Comment by aknuds1** [comment]
Also, I figure it should be safe for you to delete these empty block folders yourself (containing only the deletion marker). It looks as if Mimir is failing to do it for you.

**Reply by KevinDW-Fluxys** [reply]
ok, good to know, i will do a manual cleanup to get rid of the log spam and the empty folders

**Comment by aknuds1** [comment, marked as answer]
It could look as if your Azure Blob Storage is configured with [hierarchical namespace](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-namespace) support, could you confirm? At Grafana Labs, we are running with this option disabled, and we would suggest you do the same to avoid this problem of folders getting left behind. Mimir expects the object storage back-end to not have the folder concept, just objects with paths.

**Reply by KevinDW-Fluxys** [reply]
I can confirm that we have this enabled. However, it is not a setting you can simply disable, so that would mean we have to provision new storage and migrate. Do you think this is something that Mimr can support in the (near) future?

**Reply by aknuds1** [reply]
I don't think Mimir can support this in the near future. It would require a non-trivial rewrite of our object storage back-end logic, and it's not something we've planned at least.

**Reply by KevinDW-Fluxys** [reply]
ok, then we will migrate to a different storage account with hierarchcal namespace disabled. Once it is done i will let you know here if it solved the problem. Thanks for the quick and clear support!

**Reply by aknuds1** [reply]
Glad we could resolve it!

**Reply by KevinDW-Fluxys** [reply]
We have migrated our storage and due to the nature of the technology (no paths only files) most of it was cleaned simply by copying the files. There were some folders left with only deletion-marks which did not get cleaned up automatically, but after cleaning those manually everything is running fine and we are no longer getting any warnings in our logs.

**Comment by aknuds1** [comment]
@KevinDW-Fluxys I've made a [feature request](https://github.com/grafana/mimir/issues/5441) to track user interest in support for deleting empty object storage directories (e.g. with Azure Blob Storage hierarchical namespace).
